### ISSUE TITLE
Mask misalignment occurs during semantic segmentation inference/prediction.

### ISSUE THREAD

**Issue by 1548093028** [ultralytics/ultralytics#24617]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

When the input image resolution is 1920×1080, the preprocessing pipeline resizes the image to 640×360 and pads 12 pixels on both the top and bottom, yielding an actual model input size of 640×384. However, the semantic segmentation model produces an output of shape nc × 80 × 48. During post-processing, the padding is first computed; yet due to the 8× downsampling, the padding becomes 1.5 pixels on each side in the output feature space. After rounding, this results in 1 pixel of padding on top and 2 pixels on the bottom, ultimately causing a 0.5-pixel downward shift of the mask. When mapped back to the original image resolution, this displacement becomes 0.5 × 8 × 3 = 12 pixels, leading to a significant and noticeable mask offset.

### Environment

Ultralytics 8.4.55  Python-3.12.12 torch-2.10.0+cpu CPU (AMD RYZEN AI MAX+ 395 w/ Radeon 8060S)
Setup complete  (32 CPUs, 27.6 GB RAM, 527.5/924.5 GB disk)

OS                     Windows-11-10.0.26200-SP0
Environment            Windowstralytics.com/solutions/
Python                 3.12.12ty.ultralytics.com
Install                git.com/ultralytics/ultralytics
Path                   C:\Users\lijiq\PycharmProjects\ultralytics\ultralytics
RAM                    27.65 GB
Disk                   527.5/924.5 GBrojects\ultralytics>
CPU                    AMD RYZEN AI MAX+ 395 w/ Radeon 8060S
CPU count              32
GPU                    None
GPU count              None
CUDA                   None

numpy                   2.4.2>=1.23.0
matplotlib              3.10.8>=3.3.0
opencv-python           4.13.0.92>=4.6.0
pillow                  12.1.0>=7.1.2
pyyaml                  6.0.3>=5.3.1
requests                2.32.5>=2.23.0
scipy                   1.17.0>=1.4.1
torch                   2.10.0>=1.8.0
torch                   2.10.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision             0.25.0>=0.9.0
psutil                  7.2.2>=5.8.0
polars                  1.38.1>=0.20.0
ultralytics-thop        2.0.18>=2.0.18

### Minimal Reproducible Example

You can resize any image from Cityscapes to 1920×1080, feed it into the pretrained model for inference, and the same result should be observed.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @1548093028, thank you for reporting this possible semantic segmentation bug in Ultralytics 🚀. This is an automated response to help move things along quickly, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since this appears to be a 🐛 Bug Report, please provide a fully self-contained [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. Your current description is helpful, but a small standalone script plus one exact test image and the exact predict command or code path used would make it much easier for us to reproduce the reported mask misalignment behavior consistently.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by 1548093028** [comment]
The solution can first upsample the mask eight times, and then remove the padding. The relevant post-processing code is in ultralytics/utils/ops.py, in the scale_masks function, 、
originally

<img width="1128" height="636" alt="Image" src="https://github.com/user-attachments/assets/4e0fd399-d719-4c8c-a02d-e6cdc7130bd2" />

, after modification.
<img width="1131" height="633" alt="Image" src="https://github.com/user-attachments/assets/01aed9cc-fa7c-4aad-beff-5e3a6a0f2954" />

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24619: Fix semantic segmentation mask offset from sub-pixel LetterBox padding](https://github.com/ultralytics/ultralytics/pull/24619), thanks to @lmycross and all contributors involved.

The merged changes that may resolve this include:

- improving semantic segmentation postprocessing so predictions are aligned to the model input resolution before padding is removed
- using interpolation during postprocessing when the prediction resolution does not exactly match the input size
- converting predictions to float and reshaping them appropriately before resizing
- keeping final mask scaling back to the original image, but only after this alignment step
- reducing subtle border and padding-related mask offsets caused by sub-pixel LetterBox padding asymmetry

This should lead to cleaner and more accurate semantic segmentation masks, especially near image edges and padded regions, without requiring any API or workflow changes.

If you'd like to test the fix now, you can install the latest `main` branch with:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Or, if preferred, you can wait for the next official PyPI release.

Please let us know whether the changes in [PR #24619](https://github.com/ultralytics/ultralytics/pull/24619) resolve the issue on your side. 🙏 Thanks for reporting this, and any further feedback is welcome if the issue still persists.
